Fix false mention notifications for users in CC without mention tags

obenland opened the pull request:
## Summary

Users were incorrectly receiving mention email notifications for activities where they appeared in the CC field but were not actually mentioned in the content. This commonly occurred when users followed an actor and were automatically added to CC for federation purposes.

For example, this activity would incorrectly trigger a mention notification even though the local user is not mentioned:

```json
{
  "actor": "https://sportsbots.xyz/users/NBA",
  "object": {
    "content": "<p>Join @NateDuncanNBA and @DannyLeroux...</p>",
    "tag": [
      {"type": "Mention", "href": "https://twitter.com/NateDuncanNBA"},
      {"type": "Mention", "href": "https://twitter.com/DannyLeroux"}
    ]
  },
  "cc": ["https://obietester.blog/?author=12082291"]
}
```

The local user is in CC but not in the tags array, yet they were receiving a mention notification.

## Proposed changes:

* Modified `Mailer::mention()` to verify that a user's actor ID appears in the activity's `tag` array with `type: "Mention"` before sending a mention notification
* Previously, the function only checked if the user was in the `cc` field, which was incorrect per ActivityPub semantics

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Comprehensive test coverage has been added:
  * New test `test_mention_requires_tag_not_just_cc` verifies users in CC without mention tags don't receive notifications
  * New test `test_mention_with_tag_sends_notification` verifies users properly mentioned do receive notifications
  * Updated existing test to include proper mention tag structure
* All 28 Mailer tests pass
* Run tests with: `npm run env-test -- --filter Test_Mailer`

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Patch
- [ ] Minor
- [ ] Major

#### Type

- [ ] Added - for new features
- [ ] Changed - for changes in existing functionality
- [ ] Deprecated - for soon-to-be removed features
- [ ] Removed - for now removed features
- [x] Fixed - for any bug fixes
- [ ] Security - in case of vulnerabilities

#### Message

False mention email notifications for users in CC field without actual mention tags.

</details>

obenland committed:
Fix false mention notifications for users in CC without mention tags

Users were incorrectly receiving mention email notifications when they
appeared in the CC field of an activity but were not actually mentioned
in the content. This commonly occurred when users followed an actor and
were added to CC for federation purposes.

The fix verifies that a user's actor ID appears in the activity's tag
array with type "Mention" before sending a mention notification, rather
than just checking if they're in the CC field.

Changes:
- Add validation in Mailer::mention() to check for actual mention tags
- Add object_to_uri() support for Mention type objects
- Add test case for users in CC without mention tags (should not notify)
- Add test case for users properly mentioned (should notify)

obenland commented:
It looks like sportsbots.xyz replaces the followers URL in `cc` with individual actor IDs of their followers when they send out their activities.

This is the activity: https://browser.pub/https://sportsbots.xyz/users/FabrizioRomano/statuses/1993044518745334023

And this is what arrived in my inbox:

```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    {
      "Hashtag": "as:Hashtag",
      "sensitive": "as:sensitive",
      "dcterms": "http://purl.org/dc/terms/",
      "gts": "https://gotosocial.org/ns#",
      "interactionPolicy": {
        "@id": "gts:interactionPolicy",
        "@type": "@id"
      },
      "canQuote": {
        "@id": "gts:canQuote",
        "@type": "@id"
      },
      "canReply": {
        "@id": "gts:canReply",
        "@type": "@id"
      },
      "canLike": {
        "@id": "gts:canLike",
        "@type": "@id"
      },
      "canAnnounce": {
        "@id": "gts:canAnnounce",
        "@type": "@id"
      },
      "automaticApproval": {
        "@id": "gts:automaticApproval",
        "@type": "@id"
      },
      "manualApproval": {
        "@id": "gts:manualApproval",
        "@type": "@id"
      },
      "always": {
        "@id": "gts:always",
        "@type": "@id"
      }
    }
  ],
  "id": "https://sportsbots.xyz/users/FabrizioRomano/activities/1993044518745334023",
  "type": "Create",
  "published": "2025-11-24T19:50:16.000Z",
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "cc": [
    "https://obietester.blog/?author=12082291"
  ],
  "interactionPolicy": {
    "canQuote": {
      "automaticApproval": [
        "https://www.w3.org/ns/activitystreams#Public"
      ]
    }
  },
  "object": {
    "id": "https://sportsbots.xyz/users/FabrizioRomano/statuses/1993044518745334023",
    "type": "Note",
    "attachment": [
      {
        "type": "Document",
        "mediaType": "image/jpeg",
        "url": "https://pbs.twimg.com/media/G6i3bLMW8AAUgoD.jpg",
        "name": null,
        "blurhash": "UMDlA{NxFz%2~WShOZW=58%0$4NG57$|wvs+",
        "width": 1343,
        "height": 1343
      },
      {
        "type": "Document",
        "mediaType": "image/jpeg",
        "url": "https://pbs.twimg.com/media/G6i3bK9WIAA5Un4.jpg",
        "name": null,
        "blurhash": "UDC6DcJ~AA~BKh$kwOIqD%IoI[xbx^I.$x%f",
        "width": 544,
        "height": 544
      }
    ],
    "attributedTo": "https://sportsbots.xyz/users/FabrizioRomano",
    "content": "<p>🚨 Hansi Flick: “I’m happy to have Rashford here at Barça. I’ve always been impressed by his quality and what he can do in the box, and he’s showing that in Barcelona”.</p><p>“New football, in Barcelona, with good weather… it’s incredible to see how he’s always smiling”.</p>",
    "contentMap": {
      "en": "<p>🚨 Hansi Flick: “I’m happy to have Rashford here at Barça. I’ve always been impressed by his quality and what he can do in the box, and he’s showing that in Barcelona”.</p><p>“New football, in Barcelona, with good weather… it’s incredible to see how he’s always smiling”.</p>"
    },
    "published": "2025-11-24T19:50:16.000Z",
    "tag": [],
    "url": "https://twitter.com/FabrizioRomano/status/1993044518745334023",
    "to": [
      "https://www.w3.org/ns/activitystreams#Public"
    ],
    "cc": [
      "https://obietester.blog/?author=12082291"
    ],
    "replies": {
      "id": "https://sportsbots.xyz/replies/1993044518745334023",
      "type": "Collection",
      "first": "https://sportsbots.xyz/replies/1993044518745334023/1"
    },
    "sensitive": false,
    "interactionPolicy": {
      "canQuote": {
        "automaticApproval": [
          "https://www.w3.org/ns/activitystreams#Public"
        ]
      }
    }
  },
  "actor": "https://sportsbots.xyz/users/FabrizioRomano"
}
```

matticbot committed:
Add changelog

pfefferle commented:
Why is your blog in CC?

obenland commented:
No idea! 🙂

I assume sportsbots.xyz replaces the followers URL in `cc` with individual actor IDs of their followers when they send out their activities.

pfefferle commented on includes/class-mailer.php:
```diff
@@ -349,8 +349,13 @@ public static function mention( $activity, $user_ids ) {
+						break;
+					}
+				}
 			}
```
why even check the `cc` then? I would simply go only through the mentions instead.

obenland replied:
Sure!

obenland replied:
Fixed in [f2f08b2](https://github.com/Automattic/wordpress-activitypub/pull/2532/commits/f2f08b2e115e86eed1c23206a9814dd42397b3b6).

obenland committed:
Refactor mention handling in Mailer class

Update mention logic to use 'object.tag' for mentions instead of 'cc' recipients. Simplifies recipient mapping by filtering and mapping mention tags.

obenland committed:
Merge branch 'trunk' into fix/mention-notification-cc-only

obenland committed:
Improve readability

obenland committed:
Merge branch 'fix/mention-notification-cc-only' of https://github.com/Automattic/wordpress-activitypub into fix/mention-notification-cc-only

pfefferle commented on includes/class-mailer.php:
```diff
@@ -337,19 +337,17 @@ public static function mention( $activity, $user_ids ) {
-		// Build a map of user_id => actor_id and filter to only users in the "cc" field.
 		$recipients = array();
-		foreach ( $user_ids as $user_id ) {
+		$mentions   = wp_list_filter( (array) $activity['object']['tag'], array( 'type' => 'Mention' ) );
+		$mentions   = array_map( '\Activitypub\object_to_uri', $mentions );
```
```suggestion
		$mentions   = \wp_list_filter( (array) $activity['object']['tag'], array( 'type' => 'Mention' ) );
		$mentions   = \array_map( '\Activitypub\object_to_uri', $mentions );
```

pfefferle reviewed (approved):
Nice one! ♥️

pfefferle committed:
Merge branch 'trunk' into fix/mention-notification-cc-only

obenland committed:
Merge branch 'trunk' into fix/mention-notification-cc-only
